### ISSUE TITLE
[BUG] Multi-month date range selection shifts months shown

### ISSUE THREAD

**Issue by AbdullahShahzadd** [Hacker0x01/react-datepicker#3942]
The calendar shows 2 months (for this example I will use Jan and Feb). When I am selecting the start date in Jan and end date in Feb the months shift so that I see the months Feb and March.

What I expect to see:
<img width="513" alt="image" src="https://user-images.githubusercontent.com/77541712/219778823-03a12253-59f7-483c-8a35-ecf2578d61d2.png">
What I see:
<img width="516" alt="image" src="https://user-images.githubusercontent.com/77541712/219778987-94baa8a4-08c8-4a8c-b185-aa45d89297b2.png">

I changed the code so that `focusSelectedMonth` was dynamic and dependant on state. I made the state `true` to begin and then in `onCalendarOpen` I make it false which results in the desired behaviour. But then when I close the calendar and open it again I see Feb and March again. So that solution didn't work and was abandoned.

**To Reproduce**
```
      <DatePicker
        ref={ref}
        title={isOverflow ? dateValueString : ''}
        autoComplete='off'
        id={`date-range-${periodNumber}`}
        className={'styles'}
        // eslint-disable-next-line
        autoFocus={invalid}
        placeholderText='Start – End (m/d/yyyy)'
        selectsRange
        startDate={dateRange[0]}
        endDate={dateRange[1]}
        onChange={update => {
          onRangeChange(update);
        }}
        onChangeRaw={e => {
          if (e.target.value && e.target.value.includes('-')) {
            const dates = e.target.value.split('-').map(date => {
              const d = new Date(date);
              return isValid(d) ? d : null;
            });
            if (dates[1] > addDays(dates[0], MAX_DAYS_SELECT)) {
              dates[1] = addDays(dates[0], MAX_DAYS_SELECT);
            }
            onRangeChange(dates);
          }
        }}
        onFocus={() => {
          setOnFocusFormat(true);
        }}
        onCalendarClose={() => {
          setOnFocusFormat(false);
        }}
        dateFormat={['MM/dd/yyyy', 'MM/dd/yyyy']}
        calendarContainer={Calender}
        maxDate={maxDate}
        monthsShown={2}
        selected={dateRange[0]}
        openToDate={dateRange[0]}
        focusSelectedMonth={true}
        disabledKeyboardNavigation
        shouldCloseOnSelect={false}
        value={!onFocusFormat && dateRange[0] && dateRange[1] ? dateValueString : dateRange}
      />
```

**Expected behavior**
Added above.

**Screenshots**
Added above.

**Desktop:**

- OS: Macbook
- Browser: Chrome
- Version: 109.0.5414.119

**Comment by AbdullahShahzadd** [comment]
I have looked at 2 issues that are basically the same as this issue but they have been closed and I don't understand what the solution is:

- https://github.com/Hacker0x01/react-datepicker/issues/2609
- https://github.com/Hacker0x01/react-datepicker/issues/1337

**Comment by stubailo** [comment]
Hi! I just was looking at this issue for a long while and browsing through the code. I found this line here: https://github.com/Hacker0x01/react-datepicker/blob/8a8cb7d65186486722d829fc577cf6a158c0b8c8/src/calendar.jsx#L835

For me, setting `showPreviousMonths={false}` fixed the issue. I'm curious if it does for you too!

**Comment by stubailo** [comment]
OK I've been trying to figure this out for about an hour now and I think I've got it. The culprit is this internal state variable called [monthSelectedIn](https://github.com/Hacker0x01/react-datepicker/blob/8a8cb7d65186486722d829fc577cf6a158c0b8c8/src/calendar.jsx#L840). It represents which date the calendar thinks you want to see, and it doesn't reset when you open/close the calendar, and it's not able to be set from outside.

However, it is reset if the `monthsShown` prop is changed here: https://github.com/Hacker0x01/react-datepicker/blob/8a8cb7d65186486722d829fc577cf6a158c0b8c8/src/index.jsx#L317

So in addition to applying your hack around setting and unsetting `focusSelectedMonth`, I have added an additional hack to set and unset `monthsShown` in order to reset that internal `monthSelectedIn` state variable:

```
onCalendarClose={() => {
        setTimeout(() => setFocusSelectedMonth(true), 1);
        setTimeout(() => {
          setMonthsShownHack(1);
          setTimeout(() => setMonthsShownHack(2), 1);
        }, 1);
      }}
```

I think there's one more thing that's important here -- `getDateInView`, which sets the initial date that the calendar opens to: https://github.com/Hacker0x01/react-datepicker/blob/8a8cb7d65186486722d829fc577cf6a158c0b8c8/src/calendar.jsx#L263

This function uses the `selected` prop, but if you don't set it it does some complicated logic. So I also set `selected={startDate}`.

I think the combination of these 3 hacks results in the desired behavior, which, to recap, is:

1. You want to select a date range with 2 months shown at a time, so that it's easy to select a range that spans months
2. Clicking in the calendar when setting a time range should never cause the months displayed to change (hence turning off `focusSelectedMonth`)
3. When you close and then re-open the calendar, the first month should contain the start date of the range, and then the end date is wherever it falls. This is the best UX because the start date is the first one you're going to adjust.

Anyway, based on the hacks above the real solution is possibly resetting `monthSelectedIn` when the calendar is closed/opened. Perhaps one day I will try to open a PR.
